### PR TITLE
Build ot-cli.exe as usermode Windows executable

### PR DESCRIPTION
**Pull request by nibanks**

### CODE REVIEW

File: examples/platforms/windows/alarm.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        aTimeout->tv_sec = 10;
+        aTimeout->tv_usec = 0;
+    }
+}
+
+void windowsAlarmProcess(otInstance *aInstance)
+{
+    int32_t remaining;
+
+    if (s_is_running)
```
Comment (turon) [inline examples/platforms/windows/alarm.cpp]: Curious, is this basically a busy-loop timer?
If so, is there a more efficient facility available, such as SetTimer() if s_alarm > 100ms?

Comment (jwhui) [inline examples/platforms/windows/alarm.cpp]: @turon, this is using the select-based approach that the posix drivers implement.

Comment (jwhui) [inline examples/platforms/windows/alarm.cpp]: @turon, not a busy-loop timer. This is using the select-based approach that the posix drivers implement.

**Comment by jwhui** [comment]
Great to see windows support.  At first glance, the windows and posix drivers seem to share a lot in common.  Is it possible to share more code between the two?

**Comment by nibanks** [comment]
They do share a lot in common, except for the console handling and the socket interfaces. I could probably combine them into one source, but I don't think the directory should be called `posix` then. What do you think we should call it? Maybe something like `virtual` since it isn't on a real device?

**Comment by jwhui** [comment]
FYI, the build failure was introduced in #641.  I just merged #650 which moves the failing test back into the XFAIL list.  You'll need to rebase on top of master to get the checks to pass.

**Comment by darconeous** [comment]
I was never really fond of the name for the `posix` platform, since it doesn't really build real support for OpenThread for Posix, it builds a simulated environment.

Not sure if I like the name `virtual` better, though, but that might have more to do with my negative feelings about how the physical layer is emulated with IPv4 sockets.

I'm curious how the `posix` platform performed with the Windows Subsystem for Linux. 

I'll likely have more thoughts on this as I mull over it over the next day or so.

**Comment by nibanks** [comment]
When I build `ot-cli` for Linux it runs just find on the Windows Subsystem for Linux.

File: examples/platforms/virtual/logging.c [hunk 2]
```diff
@@ -46,21 +46,41 @@
...
+    gettimeofday(&tv, NULL);
+
+    offset = strftime(timeString, length, "%Y-%m-%d %H:%M:%S", localtime(&tv.tv_sec));
+    snprintf(&timeString[offset], length - offset, ".%06d", (uint32_t)tv.tv_usec);
+}
+#endif
+
+void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const char *aFormat, ...)
+{
+    char timeString[50];
```
Comment (darconeous) [inline examples/platforms/virtual/logging.c]: I'd be inclined to implement `gettimeofday()` wrapper on Windows and get rid of this preprocessor check.

This pattern is summed up best [here](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Generic-Functions.html):

> Autoconf follows a philosophy that was formed over the years by those who have struggled for portability: isolate the portability issues in specific files, and then program as if you were in a Posix environment. Some functions may be missing or unfixable, and your package must be ready to replace them.

However, I don't think we have any good pre-existing examples in the code base on how to do that. Will think about this some more.

**Comment by nibanks** [comment]
So what exactly do you guys want changes to accept this PR?

File: examples/platforms/virtual/platform.c [hunk 3]
```diff
@@ -80,13 +81,21 @@ void PlatformProcessDrivers(otInstance *aInstance)
...
+    }
+
     FD_ZERO(&read_fds);
     FD_ZERO(&write_fds);
     FD_ZERO(&error_fds);
 
-    posixUartUpdateFdSet(&read_fds, &write_fds, &error_fds, &max_fd);
-    posixRadioUpdateFdSet(&read_fds, &write_fds, &max_fd);
-    posixAlarmUpdateTimeout(&timeout);
+#ifndef _WIN32
```
Comment (darconeous) [inline examples/platforms/virtual/platform.c]: How about something like this instead: `#ifndef HAVE_SELECT` instead? Would need to add the following line to `configure.ac`:

```
AC_CHECK_FUNCS([select])
```

**Comment by darconeous** [comment]
It is a change that is going to cause us to have to change things internally, which tends to cause us to run behind on our own integrations. I realize that calling the platform the posix platform after this change is a bit of a misnomer, but I think it would make this change easier to swallow if we avoided renaming it for now, leaving it as "posix". Granted, it is misleading, but we have some milestones coming up and I would rather not have any renaming distractions.

Where possible, I'd rather the change use replacement implementations of missing posix functions rather than use `#ifdef`s, but I also realize that since we don't have any examples of that in this project yet that it is somewhat difficult to know what we would find acceptable for that.

I'm going to throw together a replacement implementation for `strlcpy` to replace the usage of `strncpy`. Then that should provide a model of how to do it. I'll try to punch that out today.

**Comment by nibanks** [comment]
I think my latest changes satisfy all your requests. Please let me know if there is anything else you would like changed.

**Comment by darconeous** [comment]
By the way, I put together that commit demonstrating how to add replacement functions. It is a bit larger than I originally thought, but after that commit gets merged adding new replacements for missing functions should be easy. It is tracked by #737.

File: examples/platforms/posix/platform-posix.h [hunk 4]
```diff
@@ -46,9 +46,21 @@
 #include <WinSock2.h>
 #include <WS2tcpip.h>
 #include <windows.h>
-#include <time.h>
 #define POLL WSAPoll
 #define ssize_t int
+// In user mode, define some Linux functions
+__forceinline int gettimeofday(struct timeval *tv, struct timezone *tz)
```
Comment (darconeous) [inline examples/platforms/posix/platform-posix.h]: 👍

**Pull request merged**